### PR TITLE
[TIMOB-19376]: Android fingerprint implementation

### PR DESCRIPTION
**Pull request by hieupham007**
Work in progress.

### CODE REVIEW

**Comment by ashcoding** [comment]
Trying to test this out using Emulator based on this http://developer.android.com/tools/devices/emulator.html

Not able to get it working. Would you have instructions for testing this on the latest emulator that has Fingerprint Touch Sensors included? Currently do not have a physical device with Fingerprint Sensor.

File: android/LICENSE [hunk 1]
```diff
@@ -0,0 +1,246 @@
+======================================
+Appcelerator Titanium Mobile
+Copyright 2008-2013 Appcelerator, Inc.
```
Comment (ashcoding) [inline android/LICENSE]: Probably can edit this to 2016

File: android/build.properties [hunk 2]
```diff
@@ -0,0 +1,8 @@
+titanium.sdk=/Library/Application\ Support/Titanium/
```
Comment (ashcoding) [inline android/build.properties]: We should remove ```build.properties```.

**Comment by ashcoding** [comment]
I think we don't need to include the folder ```./android/libs/``` that contains the ```*.so``` Could probably add it in ```.gitignore``` to ignore this.

File: android/src/ti/touchid/TouchidModule.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+		}
+		Activity currentActivity  = TiApplication.getInstance().getCurrentActivity();
+		if (currentActivity.checkSelfPermission(Manifest.permission.USE_FINGERPRINT) == PackageManager.PERMISSION_GRANTED) {
+			return true;
+		}
+		return false;
+	}
+
+	@Kroll.method
+	public void requestFingerprintPermissions(@Kroll.argument(optional=true)KrollFunction permissionCallback) {
```
Comment (ashcoding) [inline android/src/ti/touchid/TouchidModule.java]: Do we need this method? Based on http://developer.android.com/guide/topics/security/normal-permissions.html
```quote
If an app declares in its manifest that it needs a normal permission, the system automatically grants the app that permission at install time. The system does not prompt the user to grant normal permissions, and users cannot revoke these permissions.
```
```USE_FINGERPRINT``` is considered a normal permission as of API level 23.

**Comment by ashcoding** [comment]
Reviewed. Everything else beside what has been commented looks okay. Still not able to get the example working correctly. Any advice or instructions would be great. :)

**Comment by hieupham007** [comment]
Instructions on how to test fingerprint should be here: http://developer.android.com/about/versions/marshmallow/android-6.0.html
How are you emulating the fingerprint?

**Comment by ashcoding** [comment]
I'm using the new Android emulators that has the settings/options to emulate a fingerprint.
Info is here: http://developer.android.com/tools/devices/emulator.html

**Comment by ashcoding** [comment]
It now works for the first time the app opens.

When I click the back button and get out of the app and then get back into the app to do it again, it crashes.

Logs as follows:-
```
05-10 16:59:49.435  1540  1692 I ActivityManager: START u0 {act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] flg=0x10200000 cmp=sg.com.appcelerator.testappsg/.TestappsgActivity (has extras)} from uid 10014 on display 0
05-10 16:59:49.498  3850  3850 I TiRootActivity: (main) [0,0] checkpoint, on root activity create, savedInstanceState: null
05-10 16:59:49.526  3850  3866 F libc    : Fatal signal 11 (SIGSEGV), code 1, fault addr 0x13 in tid 3866 (KrollRuntimeThr)
05-10 16:59:49.629  1211  1211 F DEBUG   : *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
05-10 16:59:49.629  1211  1211 F DEBUG   : Build fingerprint: 'Android/sdk_google_phone_x86_64/generic_x86_64:6.0/MASTER/2780065:userdebug/test-keys'
05-10 16:59:49.629  1211  1211 F DEBUG   : Revision: '0'
05-10 16:59:49.629  1211  1211 F DEBUG   : ABI: 'x86'
05-10 16:59:49.629  1211  1211 F DEBUG   : pid: 3850, tid: 3866, name: KrollRuntimeThr  >>> sg.com.appcelerator.testappsg <<<
05-10 16:59:49.629  1211  1211 F DEBUG   : signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x13
05-10 16:59:49.643  1211  1211 F DEBUG   :     eax 00000000  ebx e2591dd4  ecx 00000000  edx 00000001
05-10 16:59:49.644  1211  1211 F DEBUG   :     esi ee7a6000  edi eab7b680
05-10 16:59:49.644  1211  1211 F DEBUG   :     xcs 00000023  xds 0000002b  xes 0000002b  xfs 0000004f  xss 0000002b
05-10 16:59:49.644  1211  1211 F DEBUG   :     eip e2115a14  ebp 00000004  esp ee77f34c  flags 00210202
05-10 16:59:49.662  1211  1211 F DEBUG   : 
05-10 16:59:49.662  1211  1211 F DEBUG   : backtrace:
05-10 16:59:49.663  1211  1211 F DEBUG   :     #00 pc 00282a14  /data/app/sg.com.appcelerator.testappsg-1/lib/x86/libkroll-v8.so
05-10 16:59:49.948  1211  1211 F DEBUG   : 
05-10 16:59:49.948  1211  1211 F DEBUG   : Tombstone written to: /data/tombstones/tombstone_04
05-10 16:59:49.948  1211  1211 E DEBUG   : AM write failed: Broken pipe
05-10 16:59:49.949  1540  1557 I BootReceiver: Copying /data/tombstones/tombstone_04 to DropBox (SYSTEM_TOMBSTONE)
05-10 16:59:49.970  1540  3889 W ActivityManager:   Force finishing activity sg.com.appcelerator.testappsg/.TestappsgActivity
05-10 16:59:50.132  1224  1224 I Zygote  : Process 3850 exited due to signal (11)
05-10 16:59:50.134  1540  1552 D GraphicsStats: Buffer count: 6
05-10 16:59:50.322  1540  3889 I WindowManager: Screenshot max retries 4 of Token{d174b0e ActivityRecord{9ac1d09 u0 sg.com.appcelerator.testappsg/.TestappsgActivity t15 f}} appWin=Window{15d6441 u0 Starting sg.com.appcelerator.testappsg} drawState=1
05-10 16:59:50.325  1540  3889 W ActivityManager: Exception thrown during pause
05-10 16:59:50.325  1540  3889 W ActivityManager: android.os.DeadObjectException
05-10 16:59:50.325  1540  3889 W ActivityManager: 	at android.os.BinderProxy.transactNative(Native Method)
05-10 16:59:50.325  1540  3889 W ActivityManager: 	at android.os.BinderProxy.transact(Binder.java:503)
05-10 16:59:50.325  1540  3889 W ActivityManager: 	at android.app.ApplicationThreadProxy.schedulePauseActivity(ApplicationThreadNative.java:727)
05-10 16:59:50.325  1540  3889 W ActivityManager: 	at com.android.server.am.ActivityStack.startPausingLocked(ActivityStack.java:867)
05-10 16:59:50.325  1540  3889 W ActivityManager: 	at com.android.server.am.ActivityStack.finishActivityLocked(ActivityStack.java:2907)
05-10 16:59:50.325  1540  3889 W ActivityManager: 	at com.android.server.am.ActivityStack.finishTopRunningActivityLocked(ActivityStack.java:2763)
05-10 16:59:50.325  1540  3889 W ActivityManager: 	at com.android.server.am.ActivityStackSupervisor.finishTopRunningActivityLocked(ActivityStackSupervisor.java:2755)
05-10 16:59:50.325  1540  3889 W ActivityManager: 	at com.android.server.am.ActivityManagerService.handleAppCrashLocked(ActivityManagerService.java:11971)
05-10 16:59:50.325  1540  3889 W ActivityManager: 	at com.android.server.am.ActivityManagerService.makeAppCrashingLocked(ActivityManagerService.java:11867)
05-10 16:59:50.325  1540  3889 W ActivityManager: 	at com.android.server.am.ActivityManagerService.crashApplication(ActivityManagerService.java:12556)
05-10 16:59:50.325  1540  3889 W ActivityManager: 	at com.android.server.am.ActivityManagerService.handleApplicationCrashInner(ActivityManagerService.java:12063)
05-10 16:59:50.325  1540  3889 W ActivityManager: 	at com.android.server.am.NativeCrashListener$NativeCrashReporter.run(NativeCrashListener.java:86)
05-10 16:59:50.326  1194  1194 E lowmemorykiller: Error opening /proc/3850/oom_score_adj; errno=2
05-10 16:59:50.370  1540  1554 W art     : Long monitor contention event with owner method=void com.android.server.am.ActivityManagerService.crashApplication(com.android.server.am.ProcessRecord, android.app.ApplicationErrorReport$CrashInfo) from ActivityManagerService.java:12502 waiters=0 for 393ms
05-10 16:59:50.374  1540  1540 W art     : Long monitor contention event with owner method=void com.android.server.am.ActivityManagerService.crashApplication(com.android.server.am.ProcessRecord, android.app.ApplicationErrorReport$CrashInfo) from ActivityManagerService.java:12502 waiters=1 for 377ms
05-10 16:59:50.376  1540  2899 I ActivityManager: Process sg.com.appcelerator.testappsg (pid 3850) has died
05-10 16:59:50.454  1540  3505 I OpenGLRenderer: Initialized EGL, version 1.4
05-10 16:59:50.651  1197  2036 E BufferQueueProducer: [Starting sg.com.appcelerator.testappsg] queueBuffer: BufferQueue has been abandoned
05-10 16:59:50.651  1540  1559 E Surface : queueBuffer: error queuing buffer to SurfaceTexture, -19
05-10 16:59:50.651  1540  1559 E Surface : queueBuffer (handle=0x7f8dedc6c1c0) failed (No such device)
05-10 16:59:50.651  1540  1559 E ViewRootImpl: Could not unlock surface
05-10 16:59:50.651  1540  1559 E ViewRootImpl: java.lang.IllegalArgumentException
05-10 16:59:50.651  1540  1559 E ViewRootImpl: 	at android.view.Surface.nativeUnlockCanvasAndPost(Native Method)
05-10 16:59:50.651  1540  1559 E ViewRootImpl: 	at android.view.Surface.unlockSwCanvasAndPost(Surface.java:301)
05-10 16:59:50.651  1540  1559 E ViewRootImpl: 	at android.view.Surface.unlockCanvasAndPost(Surface.java:282)
05-10 16:59:50.651  1540  1559 E ViewRootImpl: 	at android.view.ViewRootImpl.drawSoftware(ViewRootImpl.java:2739)
05-10 16:59:50.651  1540  1559 E ViewRootImpl: 	at android.view.ViewRootImpl.draw(ViewRootImpl.java:2642)
05-10 16:59:50.651  1540  1559 E ViewRootImpl: 	at android.view.ViewRootImpl.performDraw(ViewRootImpl.java:2434)
05-10 16:59:50.651  1540  1559 E ViewRootImpl: 	at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2067)
05-10 16:59:50.651  1540  1559 E ViewRootImpl: 	at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1107)
05-10 16:59:50.651  1540  1559 E ViewRootImpl: 	at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:6013)
05-10 16:59:50.651  1540  1559 E ViewRootImpl: 	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:858)
05-10 16:59:50.651  1540  1559 E ViewRootImpl: 	at android.view.Choreographer.doCallbacks(Choreographer.java:670)
05-10 16:59:50.651  1540  1559 E ViewRootImpl: 	at android.view.Choreographer.doFrame(Choreographer.java:606)
05-10 16:59:50.651  1540  1559 E ViewRootImpl: 	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:844)
05-10 16:59:50.651  1540  1559 E ViewRootImpl: 	at android.os.Handler.handleCallback(Handler.java:739)
05-10 16:59:50.651  1540  1559 E ViewRootImpl: 	at android.os.Handler.dispatchMessage(Handler.java:95)
05-10 16:59:50.651  1540  1559 E ViewRootImpl: 	at android.os.Looper.loop(Looper.java:148)
05-10 16:59:50.651  1540  1559 E ViewRootImpl: 	at android.os.HandlerThread.run(HandlerThread.java:61)
05-10 16:59:50.651  1540  1559 E ViewRootImpl: 	at com.android.server.ServiceThread.run(ServiceThread.java:46)
05-10 16:59:50.652  1540  1559 W WindowManager: Failed looking up window
05-10 16:59:50.652  1540  1559 W WindowManager: java.lang.IllegalArgumentException: Requested window android.view.ViewRootImpl$W@250df28 does not exist
05-10 16:59:50.652  1540  1559 W WindowManager: 	at com.android.server.wm.WindowManagerService.windowForClientLocked(WindowManagerService.java:8733)
05-10 16:59:50.652  1540  1559 W WindowManager: 	at com.android.server.wm.WindowManagerService.windowForClientLocked(WindowManagerService.java:8724)
05-10 16:59:50.652  1540  1559 W WindowManager: 	at com.android.server.wm.WindowManagerService.finishDrawingWindow(WindowManagerService.java:3449)
05-10 16:59:50.652  1540  1559 W WindowManager: 	at com.android.server.wm.Session.finishDrawing(Session.java:232)
05-10 16:59:50.652  1540  1559 W WindowManager: 	at android.view.ViewRootImpl.performDraw(ViewRootImpl.java:2472)
05-10 16:59:50.652  1540  1559 W WindowManager: 	at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2067)
05-10 16:59:50.652  1540  1559 W WindowManager: 	at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1107)
05-10 16:59:50.652  1540  1559 W WindowManager: 	at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:6013)
05-10 16:59:50.652  1540  1559 W WindowManager: 	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:858)
05-10 16:59:50.652  1540  1559 W WindowManager: 	at android.view.Choreographer.doCallbacks(Choreographer.java:670)
05-10 16:59:50.652  1540  1559 W WindowManager: 	at android.view.Choreographer.doFrame(Choreographer.java:606)
05-10 16:59:50.652  1540  1559 W WindowManager: 	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:844)
05-10 16:59:50.652  1540  1559 W WindowManager: 	at android.os.Handler.handleCallback(Handler.java:739)
05-10 16:59:50.652  1540  1559 W WindowManager: 	at android.os.Handler.dispatchMessage(Handler.java:95)
05-10 16:59:50.652  1540  1559 W WindowManager: 	at android.os.Looper.loop(Looper.java:148)
05-10 16:59:50.652  1540  1559 W WindowManager: 	at android.os.HandlerThread.run(HandlerThread.java:61)
05-10 16:59:50.652  1540  1559 W WindowManager: 	at com.android.server.ServiceThread.run(ServiceThread.java:46)
05-10 16:59:50.652  1540  1559 I Choreographer: Skipped 68 frames!  The application may be doing too much work on its main thread.
05-10 16:59:50.670  1540  1559 W WindowManager: Failed looking up window
05-10 16:59:50.670  1540  1559 W WindowManager: java.lang.IllegalArgumentException: Requested window android.view.ViewRootImpl$W@250df28 does not exist
05-10 16:59:50.670  1540  1559 W WindowManager: 	at com.android.server.wm.WindowManagerService.windowForClientLocked(WindowManagerService.java:8733)
05-10 16:59:50.670  1540  1559 W WindowManager: 	at com.android.server.wm.WindowManagerService.windowForClientLocked(WindowManagerService.java:8724)
05-10 16:59:50.670  1540  1559 W WindowManager: 	at com.android.server.wm.WindowManagerService.removeWindow(WindowManagerService.java:2697)
05-10 16:59:50.670  1540  1559 W WindowManager: 	at com.android.server.wm.Session.remove(Session.java:187)
05-10 16:59:50.670  1540  1559 W WindowManager: 	at android.view.ViewRootImpl.dispatchDetachedFromWindow(ViewRootImpl.java:3099)
05-10 16:59:50.670  1540  1559 W WindowManager: 	at android.view.ViewRootImpl.doDie(ViewRootImpl.java:5606)
05-10 16:59:50.670  1540  1559 W WindowManager: 	at android.view.ViewRootImpl$ViewRootHandler.handleMessage(ViewRootImpl.java:3409)
05-10 16:59:50.670  1540  1559 W WindowManager: 	at android.os.Handler.dispatchMessage(Handler.java:102)
05-10 16:59:50.670  1540  1559 W WindowManager: 	at android.os.Looper.loop(Looper.java:148)
05-10 16:59:50.670  1540  1559 W WindowManager: 	at android.os.HandlerThread.run(HandlerThread.java:61)
05-10 16:59:50.670  1540  1559 W WindowManager: 	at com.android.server.ServiceThread.run(ServiceThread.java:46)
```

**Comment by ashcoding** [comment]
Tried it with latest master, no issues. Bug was not due to ti.touchid.

Everything looks good. Tested with example code. 👍  

Just have questions and comments left above.
Eg. Don't have to commit ```android/libs/armeabi-v7a/libti.touchid.so```,```android/libs/armeabi/libti.touchid.so``` and ``` android/libs/x86/libti.touchid.so``` and ```build.properties```.

Also, do we need the ```hasFingerprintPermissions``` and ```requestFingerprintPermissions``` (as mentioned in my comments above)?

**Comment by hieupham007** [comment]
Addressed comments.

**Comment by ashcoding** [comment]
If you can remove `android/build.properties` as well, then I believe everything else is good to go. 👍

**Comment by ashcoding** [comment]
Looking good.
**Accepted**
Merging.

**Comment by ashcoding** [comment]
Jira: https://jira.appcelerator.org/browse/TIMOB-19376

**Pull request merged**